src/Control/Effect/Labelled.hs

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Control.Effect.Labelled
+( runLabelled
+, Labelled(Labelled)
+, LabelledMember(..)
+, HasLabelled
+, runUnderLabel
+, UnderLabel(UnderLabel)
+, module Control.Algebra
+) where
+
+import Control.Algebra
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+
+newtype Labelled (label :: k) (sub :: (* -> *) -> (* -> *)) m a = Labelled { runLabelled :: sub m a }
+  deriving (Applicative, Effect, Functor, HFunctor, Monad, MonadFail, MonadIO, MonadTrans)
+
+instance (Algebra (eff :+: sig) (sub m), HFunctor eff, HFunctor sig) => Algebra (Labelled label eff :+: sig) (Labelled label sub m) where
+  alg = \case
+    L eff -> Labelled (send (handleCoercible (runLabelled eff)))
+    R sig -> Labelled (send (handleCoercible sig))
+
+
+class LabelledMember label (sub :: (* -> *) -> (* -> *)) sup | label sup -> sub where
+  -- | Inject a member of a signature into the signature.
+  injLabelled :: Labelled label sub m a -> sup m a
```

Comment:
This class and its instances are identical to `Member`, with the exceptions of:

1. the `Labelled label` wrapper around the `sub m a`, and
2. the functional dependency which uniquely associates `label` with `sub` within `sup`.